src/PointwiseFunctions/Hydro/EquationsOfState/Tabulated3d.hpp

```diff
@@ -0,0 +1,207 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/preprocessor/arithmetic/dec.hpp>
+#include <boost/preprocessor/arithmetic/inc.hpp>
+#include <boost/preprocessor/control/expr_iif.hpp>
+#include <boost/preprocessor/list/adt.hpp>
+#include <boost/preprocessor/repetition/for.hpp>
+#include <boost/preprocessor/repetition/repeat.hpp>
+#include <boost/preprocessor/tuple/to_list.hpp>
+#include <limits>
+#include <pup.h>
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "NumericalAlgorithms/Interpolation/MultiLinearSpanInterpolation.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"  // IWYU pragma: keep
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+class DataVector;
+/// \endcond
+
+namespace EquationsOfState {
+/*!
+ * \ingroup EquationsOfStateGroup
+ * \brief Equation of state for an ideal fluid
+ *
+ * An ideal fluid equation of state:
+ *
+ * \f[
+ * p = \rho \epsilon(\gamma-1)
+ * \f]
+ *
+ * where \f$\rho\f$ is the rest mass density, \f$\epsilon\f$ is the specific
+ * internal energy, and \f$\gamma\f$ is the adiabatic index.
+ *
+ * The temperature \f$T\f$ is defined as
+ *
+ * \f[
+ * T = (\gamma - 1) \epsilon
+ * \f]
+ */
+template <bool IsRelativistic>
+class Tabulated3D : public EquationOfState<IsRelativistic, 3> {
+ public:
+  static constexpr size_t thermodynamic_dim = 3;
+  static constexpr bool is_relativistic = IsRelativistic;
+
+  struct FileName {
+    using type = std::string;
+    static constexpr Options::String help = {"Location of EoS table"};
+  };
+
+  static constexpr Options::String help = {
+      "A tabulated three-dimensional equation of state.\n"
+      "The energy density, pressure and sound speed "
+      "are tabulated as a function of density, electron_fraction and "
+      "temperature."};
+
+  using options = tmpl::list<FileName>;
+
+  /// Fields stored in the table
+  enum { EPS = 0, PRESS, CS2, NUM_TAB3D_VARS };
```

Comment:
`enum { Epsilon = 0, Pressure, CsSquared, NumberOfVars };` You can probably also declare the type, e.g. `enum : size_t` in case there are any casts needed that might be useful. What do you think about naming it? E.g. `enum Index` or `TableIndex` to give it some more context?